File: IPWorks SSH Samples/SSH Reverse Tunnel/javascript/sshreversetunnel.js
/*
 * IPWorks SSH 2022 JavaScript Edition - Sample Project
 *
 * This sample project demonstrates the usage of IPWorks SSH in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/ipworksssh
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */
 
const readline = require("readline");
const ipworksssh = require("@nsoftware/ipworksssh");

if(!ipworksssh) {
  console.error("Cannot find ipworksssh.");
  process.exit(1);
}
let rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

main();

async function main() {
  var sshreversetunnel = new ipworksssh.sshreversetunnel();

  if (process.argv.length == 2) {
    // print help/usage
    console.log(
      "Usage:\n\tnode .\\sshreversetunnel.js <user@sshserver.com> <sshpassword> <sshforwardport> [remotehost] [remoteport]"
    );
    process.exit();
  }

  sshreversetunnel.on("SSHServerAuthentication", function (e) {
    e.accept = true;
  });
  sshreversetunnel.on("Connected", function (e) {
    console.log("Connected to " + sshreversetunnel.getSSHHost() + ":" + sshreversetunnel.getSSHPort());
  });
  sshreversetunnel.on("Disconnected", function (e) {
    console.log("Disconnected");
  });
  sshreversetunnel.on("SSHChannelOpenRequest", function (e) {
    console.log("Connection from " + e.originAddress + ":" + e.originPort + " OK");
  });
  sshreversetunnel.on("SSHChannelOpened", function (e) {
    console.log("Channel[" + e.channelId + "] Opened");
  });
  sshreversetunnel.on("SSHChannelClosed", function (e) {
    console.log("Channel[" + e.channelId + "] Closed");
  });
  sshreversetunnel.on("Error", function (e) {
    console.log("Error: " + e.description);
  });

  sshreversetunnel.setLocalPort(parseInt(process.argv[2]));
  var sshinfo = process.argv[2].split("@");

  sshreversetunnel.setSSHHost(sshinfo[1]);
  sshreversetunnel.setSSHPort(22);
  sshreversetunnel.setSSHUser(sshinfo[0]);
  sshreversetunnel.setSSHPassword(process.argv[3]);
  console.log("Requesting forwarding for port " + parseInt(process.argv[4]));
  try {
    await sshreversetunnel.SSHLogon(sshinfo[1], 22);
    if (process.argv.length == 7) {
      await sshreversetunnel.requestForwarding(
        "",
        parseInt(process.argv[4]),
        process.argv[5],
        parseInt(process.argv[6])
      );
    } else {
      await sshreversetunnel.requestForwarding("", parseInt(process.argv[4]), "", 0);
    }
    console.log("Forwarding request successful.");
    while (true) {
      await sshreversetunnel.doEvents();
    }
  } catch (e) {
    if (e.code == 1098) {
      console.log("Forwarding request denied.");
    } else {
      if (e) {
        console.log(e.message);
      }
    }
  }
}

function prompt(promptName, label, punctuation, defaultVal)
{
  lastPrompt = promptName;
  lastDefault = defaultVal;
  process.stdout.write(`${label} [${defaultVal}] ${punctuation} `);
}

File: IPWorks SSH Samples/SSH Reverse Tunnel/python/sshreversetunnel.py
# 
# IPWorks SSH 2024 Python Edition - Sample Project
# 
# This sample project demonstrates the usage of IPWorks SSH in a 
# simple, straightforward way. It is not intended to be a complete 
# application. Error handling and other checks are simplified for clarity.
# 
# www.nsoftware.com/ipworksssh
# 
# This code is subject to the terms and conditions specified in the 
# corresponding product license agreement which outlines the authorized 
# usage and restrictions.
# 

import sys
import string
from ipworksssh import *

input = sys.hexversion<0x03000000 and raw_input or input

def ensureArg(args, prompt, index):
  if len(args) <= index:
    while len(args) <= index:
      args.append(None)
    args[index] = input(prompt)
  elif args[index] is None:
    args[index] = input(prompt)



def fireConnected(e):
    print("Connected to server")

def fireError(e):
  print("StdErr: %s" % e.description)

def fireLog(e):
  print("Log: %s" % e.message)

def fireSSHServerAuth(e):
  print("\nServer provided the following fingerprint:\n%s\n"%e.fingerprint)
  cntue = input("Would you like to continue [y/n]: ")
  if cntue == "Y" or cntue == "y":
      e.accept = True

def setupHost():
  host = input("SSH Host: ")
  sshreversetunnel.ssh_host = host
  port = input("SSH Port [22]: ")
  if (port == ""):
      port = 22
  sshreversetunnel.ssh_port = int(port)
  user = input("SSH User: ")
  sshreversetunnel.ssh_user = user
  password = input("SSH Password: ")
  sshreversetunnel.ssh_password = password

sshreversetunnel = SSHReverseTunnel()
sshreversetunnel.on_connected = fireConnected
sshreversetunnel.on_ssh_server_authentication = fireSSHServerAuth
sshreversetunnel.on_log = fireLog
sshreversetunnel.on_error = fireError

try:
    setupHost()
    sshreversetunnel.ssh_logon(sshreversetunnel.ssh_host, sshreversetunnel.ssh_port)

    sshforwardport = input("Forward port [4444]: ")
    if (sshforwardport == ""):
        sshforwardport = "4444"
    remotehost = input("Remote host [www.nsoftware.com]: ")
    if (remotehost == ""):
        remotehost = "www.nsoftware.com"
    remoteport = input("Remote port [443]: ")
    if (remoteport == ""):
        remoteport = "443"

    print("Requesting forwarding for port " + sshforwardport + "; forwarding to: " + remotehost + ":" + remoteport)
    sshreversetunnel.request_forwarding("0.0.0.0", int(sshforwardport), remotehost, int(remoteport))
    print("Forwarding request successful.")
    print("Press Ctrl-C to quit...")
    try:
      while True:
        sshreversetunnel.do_events()
    except KeyboardInterrupt:
      print("Exiting...")
      sshreversetunnel.ssh_logoff()
except IPWorksSSHError as e:
    print( "Error %s"%e.message)
except KeyboardInterrupt:
    print( "Exiting..." )


